kayenta-prometheus/src/main/java/com/netflix/kayenta/prometheus/metrics/PrometheusMetricDescriptorsCache.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2017 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.kayenta.prometheus.metrics;
+
+import com.netflix.kayenta.prometheus.model.PrometheusMetricDescriptor;
+import com.netflix.kayenta.prometheus.model.PrometheusMetricDescriptorsResponse;
+import com.netflix.kayenta.prometheus.security.PrometheusNamedAccountCredentials;
+import com.netflix.kayenta.prometheus.service.PrometheusRemoteService;
+import com.netflix.kayenta.security.AccountCredentials;
+import com.netflix.kayenta.security.AccountCredentialsRepository;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class PrometheusMetricDescriptorsCache {
+
+    private volatile Map<String, List<PrometheusMetricDescriptor>> cache = Collections.emptyMap();
```

Comment:
this was not volatile before, but it can be updated in the different thread, so we do want to guarantee that all threads receive the changes.